File: src/main/java/net/Chidoziealways/everythingjapanese/jutsu/Jutsu.java
package net.Chidoziealways.everythingjapanese.jutsu;

import com.mojang.serialization.DataResult;
import net.minecraft.core.component.DataComponentMap;
import net.minecraft.core.component.DataComponentType;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Unit;
import net.minecraft.world.entity.player.Player;

import javax.annotation.Nullable;
import java.util.Set;

public abstract class Jutsu {
    private final ResourceLocation id; // Unique identifier for the Jutsu
    protected final String name;         // Name of the Jutsu (localized)
    private final int chakraCost;      // Chakra cost to use the Jutsu
    private final int powerLevel;      // Power level or damage output
    private final JutsuType[] type;      // Type of Jutsu (Ninjutsu, Genjutsu, Taijutsu, etc.)

    public Jutsu(ResourceLocation id, String name, int chakraCost, int powerLevel, JutsuType... type) {
        this.id = id;
        this.name = name;
        this.chakraCost = chakraCost;
        this.powerLevel = powerLevel;
        this.type = type;
    }

    public static DataResult<Unit> validateComponents(DataComponentMap components) {
        return DataResult.success(Unit.INSTANCE);
    }

    @Nullable
    private DataComponentMap builtComponents = null;

    public DataComponentMap components() {
        if (builtComponents == null) {
            builtComponents = new DataComponentMap() {
                @Override
                public Set<DataComponentType<?>> keySet() {
                    return Set.of();
                }

                @Override
                public @org.jetbrains.annotations.Nullable <T> T get(DataComponentType<? extends T> p_395766_) {
                    return null;
                }
            };
        }

        return builtComponents;
    }

    public ResourceLocation getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getChakraCost() {
        return chakraCost;
    }

    public int getPowerLevel() {
        return powerLevel;
    }



    public JutsuType[] getType() {
        return type;
    }

    @Override
    public String toString() {
        return "Jutsu{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", chakraCost=" + chakraCost +
                ", powerLevel=" + powerLevel +
                ", type=" + type +
                '}';
    }

    public abstract void cast(Player player);
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/jutsu/Jutsu.kt
package net.Chidoziealways.everythingjapanese.jutsu

import com.mojang.serialization.DataResult
import net.minecraft.core.component.DataComponentMap
import net.minecraft.core.component.DataComponentType
import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerPlayer
import net.minecraft.util.Unit
import net.minecraft.world.entity.player.Player

abstract class Jutsu(// Unique identifier for the Jutsu
    private val id: Identifier, // Name of the Jutsu (localized)
    val name: String?, // Chakra cost to use the Jutsu
    private val chakraCost: Float, // Power level or damage output
    private val powerLevel: Int, vararg type: JutsuType?
) {
    private val type: Array<JutsuType?>? =
        type as Array<JutsuType?>? // Type of Jutsu (Ninjutsu, Genjutsu, Taijutsu, etc.)

    private var builtComponents: DataComponentMap? = null

    fun components(): DataComponentMap {
        if (builtComponents == null) {
            builtComponents = object : DataComponentMap {
                override fun keySet(): MutableSet<DataComponentType<*>> {
                    return mutableSetOf<DataComponentType<*>>()
                }

                override fun <T : Any> get(p_395766_: DataComponentType<out T>): T? {
                    return null
                }
            }
        }

        return builtComponents!!
    }

    fun getID(): Identifier {
        return id
    }

    fun getChakraCost(): Float {
        return chakraCost
    }

    fun getPowerLevel(): Int {
        return powerLevel
    }


    fun getType(): Array<JutsuType?>? {
        return type
    }

    override fun toString(): String {
        return "Jutsu{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", chakraCost=" + chakraCost +
                ", powerLevel=" + powerLevel +
                ", type=" + type +
                '}'
    }

    abstract fun cast(player: ServerPlayer): Boolean

    companion object {
        fun validateComponents(components: DataComponentMap): DataResult<Unit> {
            return DataResult.success<Unit?>(Unit.INSTANCE)
        }
    }
}
